src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1577,57 +1577,56 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Sum")>]
         let inline sum (source: seq< (^a) >) : ^a = 
-            use e = source.GetEnumerator() 
             let mutable acc = LanguagePrimitives.GenericZero< (^a) >
-            while e.MoveNext() do
-                acc <- Checked.(+) acc e.Current
+            for item in source do
+                acc <- Checked.(+) acc item
```

Comment:
We could in theory do that. 

One problem is that in the case of struct-enumerators, we don't have a guarantee that the struct-version is semantically equivalent to the interface version.  So an optimization like this is not 100% sound - it's a heuristic but one that's very likely to be sound - and for arrays it would definitely be sound.